Allow checksum file to be local file:// url

madeddie opened the pull request:
##### SUMMARY
A partial fix for #69364 in that the SHASUMS file can be downloaded and GPG verified but then used from the downloaded location to verify the get_url's file.

or
When using get_url to download a file that has a GPG signed SHASUMS file, this would allow for less filter hackary.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
get_url

madeddie committed:
Allow checksum file to be local file:// url

This would be a partial solution for #69364 in that the SHASUMS file can be downloaded and gpg verified but then used from the downloaded location to verify the get_url's file.

madeddie commented:
Not quite sure why shippable is failing, it looks like a transient error in CI, but if there's a specific error I need to look at, please let me know. Testing succeeds on my local system.

madeddie commented on lib/ansible/modules/get_url.py:
```diff
@@ -487,7 +487,7 @@ def main():
             module.fail_json(msg="The checksum parameter has to be in format <algorithm>:<checksum>", **result)
 
-        if checksum.startswith('http://') or checksum.startswith('https://') or checksum.startswith('ftp://'):
+        if checksum.startswith('http://') or checksum.startswith('https://') or checksum.startswith('ftp://') or checksum.startswith('file://'):
```
Maybe shortening this to
`if checksum.startswith(('http://', 'https://', 'ftp://', 'file://')):`
might be nice too.

samdoran replied:
Yes, please use the suggestion above to shorten this statement.

bcoca replied:
just making note for future, not holding this PR to it.

in the end we should really use parse_url and compare the 'schema' to 'supported' schema handlers, which startswith is 'kindof' doing here but it would be cleaner this way. Also would require to 'register' supported handlers and possibly giving a way to expand upon them.

madeddie replied:
Made a function using urlsplit and using that. Possibility to expanding I'll leave to a later exercise.

bmillemathias commented:
/rebuild_failed

bmillemathias commented:
Hi, thanks for the PR
> Not quite sure why shippable is failing, it looks like a transient error in CI, but if there's a specific error I need to look at, please let me know. Testing succeeds on my local system.

I've launched again the CI.

To complete the PR, that would be great to add a test with a file:// so we can make sure the code works and won't break, and also add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#creating-a-changelog-fragment).

Best

samdoran commented:
Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and integration tests. See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

madeddie committed:
Make checksum url parsing more explicit

Use urlsplit to test if the checksum string has a (currently tested and) supported url scheme.

madeddie committed:
Fix whitespace

madeddie committed:
Changelog fragment

madeddie committed:
Added tests

Commit:
Fix typo in test setup

samdoran commented:
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs) against `stable-2.9` and `stable-2.10` for this to be included in previous versions.

Akasurde commented:
As of 18 June, 2021 - Stable-2.9 is for critical fixes. 
Backport for Stable-2.10 - https://github.com/ansible/ansible/pull/75052
